File: src/main/java/com/wesker926/leetcode/algorithms/p0851/s1/Solution.java
package com.wesker926.leetcode.algorithms.p0851.s1;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

/**
 * @author wesker.gh
 * @date 2021/12/15
 * @description 构造图 + dfs
 */
public class Solution {

    private int[] quiet;

    private int[] ans;

    private List[] relation;

    public int[] loudAndRich(int[][] richer, int[] quiet) {
        int n = quiet.length;
        relation = new List[n];
        for (int i = 0; i < n; i++) {
            relation[i] = new LinkedList<Integer>();
        }
        for (int[] ints : richer) {
            relation[ints[1]].add(ints[0]);
        }

        this.quiet = quiet;
        ans = new int[n];
        Arrays.fill(ans, -1);

        for (int i = 0; i < n; i++) {
            dfs(i);
        }
        return ans;
    }

    private void dfs(int i) {
        if (ans[i] != -1) {
            return;
        }

        ans[i] = i;
        for (Object o : relation[i]) {
            int j = (int) (o);
            dfs(j);
            if (quiet[ans[i]] > quiet[ans[j]]) {
                ans[i] = ans[j];
            }
        }
    }
}

File: src/main/java/com/wesker926/leetcode/algorithms/p0851/s1/solution.go
// Package s1
// @author: wesker.gh
// @date: 2021/12/15
// @description: 构造图 + dfs
package s1

var (
	q        []int
	ans      []int
	relation [][]int
)

func loudAndRich(richer [][]int, quiet []int) []int {
	n := len(quiet)
	relation = make([][]int, n)
	for _, ints := range richer {
		relation[ints[1]] = append(relation[ints[1]], ints[0])
	}

	q, ans = quiet, make([]int, n)
	for i := 0; i < n; i++ {
		ans[i] = n
	}

	for i := 0; i < n; i++ {
		dfs(i)
	}
	return ans
}

func dfs(i int) {
	if ans[i] != len(q) {
		return
	}

	ans[i] = i
	for _, j := range relation[i] {
		dfs(j)
		if q[ans[i]] > q[ans[j]] {
			ans[i] = ans[j]
		}
	}
}
